File: ObjCStarter/ObjCStarter/CustomDomainViewController.m
#import "CustomDomainViewController.h"
#import <CleverTapSDK/CleverTap.h>
#import <CleverTapSDK/CleverTapInstanceConfig.h>
#import <CleverTapSDK/CleverTap+Inbox.h>

@interface CustomDomainViewController() <UITableViewDelegate, UITableViewDataSource, CleverTapInboxViewControllerDelegate>

@property (weak, nonatomic) IBOutlet UITableView *tblView;
@property (nonatomic, strong) NSMutableArray *eventList;
@property (nonatomic, strong) CleverTap *cleverTapAdditionalInstance;
@end

@implementation CustomDomainViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    // Do any additional setup after loading the view.
    CleverTapInstanceConfig *ctConfig = [[CleverTapInstanceConfig alloc] initWithAccountId:@"R65-RR9-9R5Z" accountToken:@"c22-562" proxyDomain:@"analytics.sdktesting.xyz"];
//    or to use spiky proxy domain call
//    [[CleverTapInstanceConfig alloc] initWithAccountId:@"R65-RR9-9R5Z" accountToken:@"c22-562" proxyDomain:@"analytics.sdktesting.xyz" spikyProxyDomain:@"analyticst.sdktesting.xyz"];
    [ctConfig setLogLevel: CleverTapLogDebug];
    _cleverTapAdditionalInstance = [CleverTap instanceWithConfig:ctConfig];
    
    [self loadData];
    [self initializeAppInbox];
    self.tblView.tableFooterView = [UIView new];
}

- (void)loadData {
    [_cleverTapAdditionalInstance recordScreenView:@"CustomDomainViewController"];
    self.eventList = [[NSMutableArray alloc] initWithObjects:@"Record User Profile",
                      @"Record User Profile with Properties",
                      @"Record User Event called Product Viewed",
                      @"Record User Event with Properties",
                      @"Record User Charged Event",
                      @"Open App Inbox",
                      @"Increment User Profile Property",
                      @"Decrement User Profile Property", nil];
    [self.tblView reloadData];
}
    
- (void)initializeAppInbox {
    [_cleverTapAdditionalInstance initializeInboxWithCallback:^(BOOL success) {
        int messageCount = (int)[self.cleverTapAdditionalInstance getInboxMessageCount];
        int unreadCount = (int)[self.cleverTapAdditionalInstance getInboxMessageUnreadCount];
        NSLog(@"Inbox Message: %d/%d", messageCount, unreadCount);
    }];
}

#pragma mark - Table view Delegate and Data Source

- (nonnull UITableViewCell *)tableView:(nonnull UITableView *)tableView cellForRowAtIndexPath:(nonnull NSIndexPath *)indexPath {
    
    static NSString *simpleTableIdentifier = @"cell";
    UITableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:simpleTableIdentifier];
    
    if (cell == nil) {
        cell = [[UITableViewCell alloc] initWithStyle:UITableViewCellStyleDefault reuseIdentifier:simpleTableIdentifier];
    }
    
    cell.textLabel.text = [self.eventList objectAtIndex:indexPath.row];
    cell.accessoryType = UITableViewCellAccessoryDisclosureIndicator;
    return cell;
}

- (NSInteger)tableView:(nonnull UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return _eventList.count;
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath
{
    switch (indexPath.row) {
        case 0:
            [self recordUserProfile];
            break;
        case 1:
            [self recordUserProfileWithProperties];
            break;
        case 2:
            [self recordUserEventWithoutProperties];
            break;
        case 3:
            [self recordUserEventWithProperties];
            break;
        case 4:
            [self recordUserChargedEvent];
            break;
        case 5:
            [self showAppInbox];
            break;
        case 6:
            [self incrementUserProfileProperty];
            break;
        case 7:
            [self decrementUserProfileProperty];
            break;
        default:
            break;
    }
    [tableView deselectRowAtIndexPath: indexPath animated: YES];
}

- (void)recordUserProfile {
    
    // each of the below mentioned fields are optional
    // if set, these populate demographic information in the Dashboard
    NSDateComponents *dob = [[NSDateComponents alloc] init];
    dob.day = 24;
    dob.month = 5;
    dob.year = 1992;
    NSDate *d = [[NSCalendar currentCalendar] dateFromComponents:dob];
    NSDictionary *profile = @{
                              @"Name": @"Jack Montana",               // String
                              @"Identity": @61026032,                 // String or number
                              @"Email": @"jack@gmail.com",            // Email address of the user
                              @"Phone": @"+14155551234",              // Phone (with the country code, starting with +)
                              @"Gender": @"M",                        // Can be either M or F
                              @"Employed": @"Y",                      // Can be either Y or N
                              @"Education": @"Graduate",              // Can be either Graduate, College or School
                              @"Married": @"Y",                       // Can be either Y or N
                              @"DOB": d,                              // Date of Birth. An NSDate object
                              @"Age": @28,                            // Not required if DOB is set
                              @"Tz": @"Asia/Kolkata",                 //an abbreviation such as "PST", a full name such as "America/Los_Angeles",
                              //or a custom ID such as "GMT-8:00"
                              @"Photo": @"www.foobar.com/image.jpeg", // URL to the Image
                              
                              // optional fields. controls whether the user will be sent email, push etc.
                              @"MSG-email": @NO,                      // Disable email notifications
                              @"MSG-push": @YES,                      // Enable push notifications
                              @"MSG-sms": @NO,                         // Disable SMS notifications
                              
                              //custom fields
                              @"score": @15,
                              @"cost": @10.5

                              };
    
    [_cleverTapAdditionalInstance profilePush:profile];
}
- (void)recordUserProfileWithProperties {
    // To set a multi-value property
    [_cleverTapAdditionalInstance profileSetMultiValues:@[@"bag", @"shoes"] forKey:@"myStuff"];
    
    // To add an additional value(s) to a multi-value property
    [_cleverTapAdditionalInstance profileAddMultiValue:@"coat" forKey:@"myStuff"];
    // or
    [_cleverTapAdditionalInstance profileAddMultiValues:@[@"socks", @"scarf"] forKey:@"myStuff"];
    
    //To remove a value(s) from a multi-value property
    [_cleverTapAdditionalInstance profileRemoveMultiValue:@"bag" forKey:@"myStuff"];
    [_cleverTapAdditionalInstance profileRemoveMultiValues:@[@"shoes", @"coat"] forKey:@"myStuff"];
    
    //To remove the value of a property (scalar or multi-value)
    [_cleverTapAdditionalInstance profileRemoveValueForKey:@"myStuff"];
}
- (void)recordUserEventWithoutProperties {
    // event without properties
    [_cleverTapAdditionalInstance recordEvent:@"Product viewed"];
}
- (void)recordUserEventWithProperties {
    // event with properties
    NSDictionary *props = @{
                            @"Product name": @"Casio Chronograph Watch",
                            @"Category": @"Mens Accessories",
                            @"Price": @59.99,
                            @"Date": [NSDate date]
                            };
    
    [_cleverTapAdditionalInstance recordEvent:@"Product viewed" withProps:props];
}
- (void)recordUserChargedEvent {
    NSDictionary *chargeDetails = @{
                                    @"Amount" : @300,
                                    @"Payment mode": @"Credit Card",
                                    @"Charged ID": @24052013
                                    };
    
    NSDictionary *item1 = @{
                            @"Category": @"books",
                            @"Book name": @"The Millionaire next door",
                            @"Quantity": @1
                            };
    
    NSDictionary *item2 = @{
                            @"Category": @"books",
                            @"Book name": @"Achieving inner zen",
                            @"Quantity": @1
                            };
    
    NSDictionary *item3 = @{
                            @"Category": @"books",
                            @"Book name": @"Chuck it, let's do it",
                            @"Quantity": @5
                            };
    
    NSArray *items = @[item1, item2, item3];
    [_cleverTapAdditionalInstance recordChargedEventWithDetails:chargeDetails
                                                     andItems:items];
}

- (void)showAppInbox {
    
    CleverTapInboxStyleConfig *style = [[CleverTapInboxStyleConfig alloc] init];
    style.messageTags = @[@"tag1", @"tag2"];
    style.tabSelectedBgColor = [UIColor blueColor];
    style.tabSelectedTextColor = [UIColor whiteColor];
    CleverTapInboxViewController *inboxController = [_cleverTapAdditionalInstance newInboxViewControllerWithConfig:style andDelegate:self];
    if (inboxController) {
        UINavigationController *navigationController = [[UINavigationController alloc] initWithRootViewController:inboxController];
        [self presentViewController:navigationController animated:YES completion:nil];
    }
}
- (void)messageDidSelect:(CleverTapInboxMessage *)message atIndex:(int)index withButtonIndex:(int)buttonIndex {
    //  This is called when an inbox message is clicked(tapped or call to action)
}

- (void)incrementUserProfileProperty {
    [_cleverTapAdditionalInstance profileIncrementValueBy: @3 forKey: @"score"];
}

- (void)decrementUserProfileProperty {
    [_cleverTapAdditionalInstance profileDecrementValueBy: @7 forKey: @"score"];
}

- (void)activateCustomDomain {
    UIStoryboard *storyBoard = [UIStoryboard storyboardWithName:@"Main" bundle:nil];
    UIViewController *customDomainVC = [storyBoard instantiateViewControllerWithIdentifier:@"CustomDomainVC"];
    [self.navigationController pushViewController:customDomainVC animated:YES];
}

@end

File: SwiftStarter/SwiftStarter/CustomDomainViewController.swift
import UIKit
import CleverTapSDK

class CustomDomainViewController: UIViewController, CleverTapInboxViewControllerDelegate {
    
    @IBOutlet var tblEvent: UITableView!
    var eventList: [String] = [String]()
    
    lazy var cleverTapAdditionalInstance: CleverTap = {
        let ctConfig = CleverTapInstanceConfig(accountId: "R65-RR9-9R5Z", accountToken: "c22-562", proxyDomain: "analytics.sdktesting.xyz")
//        or to use spiky proxy domain call
//        CleverTapInstanceConfig(accountId: "W9R-486-4W5Z", accountToken: "6b4-2c0", proxyDomain: "analytics.sdktesting.xyz", spikyProxyDomain: "analyticst.sdktesting.xyz")
        return CleverTap.instance(with: ctConfig)
    }()
    
    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view, typically from a nib.
        cleverTapAdditionalInstance.recordScreenView("CustomDomainViewController")
        loadData()
        registerAppInbox()
        initializeAppInbox()
        tblEvent.tableFooterView = UIView()
    }
    
    func messageDidSelect(_ message: CleverTapInboxMessage, at index: Int32, withButtonIndex buttonIndex: Int32) {
        //  This is called when an inbox message is clicked(tapped or call to action)
    }
}

extension CustomDomainViewController {
    
    func loadData(){
        eventList.append("Record User Profile")
        eventList.append("Record User Profile with Properties")
        eventList.append("Record User Event called Product Viewed")
        eventList.append("Record User Event with Properties")
        eventList.append("Record User Charged Event")
        eventList.append("Show App Inbox")
        eventList.append("Increment User Profile Property")
        eventList.append("Decrement User Profile Property")
        self.tblEvent.reloadData()
    }
    
    func registerAppInbox() {
        cleverTapAdditionalInstance.registerInboxUpdatedBlock({ [weak self] in
            let messageCount = self?.cleverTapAdditionalInstance.getInboxMessageCount()
            let unreadCount = self?.cleverTapAdditionalInstance.getInboxMessageUnreadCount()
            print("Inbox Message:\(String(describing: messageCount))/\(String(describing: unreadCount)) unread")
        })
    }
    
    func initializeAppInbox() {
        cleverTapAdditionalInstance.initializeInbox(callback: ({ [weak self] (success) in
            let messageCount = self?.cleverTapAdditionalInstance.getInboxMessageCount()
            let unreadCount = self?.cleverTapAdditionalInstance.getInboxMessageUnreadCount()
            print("Inbox Message:\(String(describing: messageCount))/\(String(describing: unreadCount)) unread")
        }))
    }
}
extension CustomDomainViewController: UITableViewDataSource, UITableViewDelegate{
    
    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        return self.eventList.count
    }
    
    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        let cell: UITableViewCell = tableView.dequeueReusableCell(withIdentifier: "cell")!
        cell.textLabel?.text = self.eventList[indexPath.row]
        cell.accessoryType = UITableViewCell.AccessoryType.disclosureIndicator
        return cell
    }
    
    func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        print("You selected cell #\(indexPath.row)!")
        
        switch(indexPath.row)
        {
        case 0: recordUserProfile()
        case 1: recordUserProfileWithProperties()
        case 2: recordUserEventWithoutProperties()
        case 3: recordUserEventWithProperties()
        case 4: recordUserChargedEvent()
        case 5: showAppInbox()
        case 6: incrementUserProfileProperty()
        case 7: decrementUserProfileProperty()
        default: break;
        }
        tableView.deselectRow(at: indexPath, animated: true)
    }
    
    func recordUserProfile() {
        
        // each of the below mentioned fields are optional
        // if set, these populate demographic information in the Dashboard
        let dob = NSDateComponents()
        dob.day = 24
        dob.month = 5
        dob.year = 1992
        let d = NSCalendar.current.date(from: dob as DateComponents)
        let profile: Dictionary<String, AnyObject> = [
            "Name": "Jack Montana" as AnyObject,                 // String
            "Identity": 61026032 as AnyObject,                   // String or number
            "Email": "jack@gmail.com" as AnyObject,              // Email address of the user
            "Phone": "+14155551234" as AnyObject,                // Phone (with the country code, starting with +)
            "Gender": "M" as AnyObject,                          // Can be either M or F
            "Employed": "Y" as AnyObject,                        // Can be either Y or N
            "Education": "Graduate" as AnyObject,                // Can be either School, College or Graduate
            "Married": "Y" as AnyObject,                         // Can be either Y or N
            "DOB": d! as AnyObject,                              // Date of Birth. An NSDate object
            "Age": 26 as AnyObject,                              // Not required if DOB is set
            "Tz":"Asia/Kolkata" as AnyObject,                    //an abbreviation such as "PST", a full name such as "America/Los_Angeles",
            //or a custom ID such as "GMT-8:00"
            "Photo": "www.foobar.com/image.jpeg" as AnyObject,   // URL to the Image

            // optional fields. controls whether the user will be sent email, push etc.
            "MSG-email": false as AnyObject,                     // Disable email notifications
            "MSG-push": true as AnyObject,                       // Enable push notifications
            "MSG-sms": false as AnyObject,                        // Disable SMS notifications

            //custom fields
            "score": 15 as AnyObject,
            "cost": 10.5 as AnyObject
        ]
        
        cleverTapAdditionalInstance.profilePush(profile)
    }
    
    func recordUserProfileWithProperties() {
        // To set a multi-value property
        cleverTapAdditionalInstance.profileSetMultiValues(["bag", "shoes"], forKey: "myStuff")
        
        // To add an additional value(s) to a multi-value property
        cleverTapAdditionalInstance.profileAddMultiValue("coat", forKey: "myStuff")
        // or
        cleverTapAdditionalInstance.profileAddMultiValues(["socks", "scarf"], forKey: "myStuff")
        
        //To remove a value(s) from a multi-value property
        cleverTapAdditionalInstance.profileRemoveMultiValue("bag", forKey: "myStuff")
        cleverTapAdditionalInstance.profileRemoveMultiValues(["shoes", "coat"], forKey: "myStuff")

        //To remove the value of a property (scalar or multi-value)
        cleverTapAdditionalInstance.profileRemoveValue(forKey: "myStuff")
    }
    
    func recordUserEventWithoutProperties() {
        // event without properties
        cleverTapAdditionalInstance.recordEvent("Product viewed")
    }
    
    func recordUserEventWithProperties() {
        // event with properties
        let props = [
            "Product name": "Casio Chronograph Watch",
            "Category": "Mens Accessories",
            "Price": 59.99,
            "Date": NSDate()
            ] as [String : Any]
        cleverTapAdditionalInstance.recordEvent("Product viewed", withProps: props)
    }
    
    func recordUserChargedEvent() {
        let chargeDetails = [
            "Amount": 300,
            "Payment mode": "Credit Card",
            "Charged ID": 24052013
            ] as [String : Any]
        
        let item1 = [
            "Category": "books",
            "Book name": "The Millionaire next door",
            "Quantity": 1
            ] as [String : Any]
        
        let item2 = [
            "Category": "books",
            "Book name": "Achieving inner zen",
            "Quantity": 1
            ] as [String : Any]
        
        let item3 = [
            "Category": "books",
            "Book name": "Chuck it, let's do it",
            "Quantity": 5
            ] as [String : Any]
        
        cleverTapAdditionalInstance.recordChargedEvent(withDetails: chargeDetails, andItems: [item1, item2, item3])
    }
    
    func showAppInbox() {
        
        // config the style of App Inbox Controller
        let style = CleverTapInboxStyleConfig.init()
        style.title = "App Inbox"
        style.navigationTintColor = .black
        
        if let inboxController = cleverTapAdditionalInstance.newInboxViewController(with: style, andDelegate: self) {
            let navigationController = UINavigationController.init(rootViewController: inboxController)
            self.present(navigationController, animated: true, completion: nil)
        }
    }
    
    func incrementUserProfileProperty() {
        cleverTapAdditionalInstance.profileIncrementValue(by: NSNumber(value: 1), forKey: "score")
    }
    
    func decrementUserProfileProperty() {
        cleverTapAdditionalInstance.profileDecrementValue(by: NSNumber(value: 1), forKey: "score")
    }
}

